renku/core/management/config.py

```diff
@@ -35,97 +35,127 @@ def print_app_config_path(ctx, param, value):
 
-    def store_config(self, config):
-        """Persists global configuration object."""
-        os.umask(0)
-        fd = os.open(
-            str(self.config_path), os.O_CREAT | os.O_RDWR | os.O_TRUNC, 0o600
-        )
+        if not local_only:
+            with _GlobalConfigLock(self):
+                config.read(config_files)
+        else:
+            config.read(config_files)
+        return config
 
-        with open(fd, 'w+') as file:
-            config.write(file)
+    def store_config(self, config, global_only):
+        """Persists locally or globally configuration object.
+
+        Global configuration is updated only when :global_only: is True,
+        otherwise, updates are written to local project configuration
+        """
+        filepath = self.global_config_path if global_only else \
+            self.local_config_path
+
+        if global_only:
+            os.umask(0)
+            fd = os.open(filepath, os.O_CREAT | os.O_RDWR | os.O_TRUNC, 0o600)
+            with _GlobalConfigLock(self):
+                with open(fd, 'w+') as file:
+                    config.write(file)
+        else:
+            with open(filepath, 'w+') as file:
+                config.write(file)
 
-        return self.load_config()
+        return self.load_config(local_only=True, global_only=True)
 
-    def get_value(self, section, key):
+    def get_value(self, section, key, local_only=False, global_only=False):
         """Get value from specified section and key."""
-        config = self.load_config()
+        config = self.load_config(
+            local_only=local_only, global_only=global_only
+        )
         return config.get(section, key, fallback=None)
 
-    def set_value(self, section, key, value):
+    def set_value(self, section, key, value, global_only=False):
         """Set value to specified section and key."""
-        config = self.load_config()
+        local_only = not global_only
+        config = self.load_config(
+            local_only=local_only, global_only=global_only
+        )
         if section in config:
             config[section][key] = value
         else:
             config[section] = {key: value}
 
-        config = self.store_config(config)
-        return config
+        self.store_config(config, global_only=global_only)
 
-    def remove_value(self, section, key):
+    def remove_value(self, section, key, global_only=False):
         """Remove key from specified section."""
-        config = self.load_config()
+        local_only = not global_only
+        config = self.load_config(
+            local_only=local_only, global_only=global_only
+        )
 
         if section in config:
             config[section].pop(key)
 
             if not config[section].keys():
                 config.pop(section)
 
-        config = self.store_config(config)
-        return config
+            self.store_config(config, global_only=global_only)
+
+
+class _GlobalConfigLock:
```

Comment:
The convention in the repository to achieve this is usually by using context managers.

Reply:
This is a context manager but I moved it to a different class. Do you mean a context manager in the same class (i.e. LocalClient)?

Reply:
This [way](https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager) is used in the repo. Also, we should write the handling of the lock files in the similar way which exists already in the [repo](https://github.com/SwissDataScienceCenter/renku-python/blob/master/renku/core/management/repository.py#L115).

Reply:
That was way cleaner :sweat_smile: